Define encoding of PHP scripts (Fix #233)

Agares opened the pull request:
1. Remove references to unicode codepoints (as those were ambigouous
without a defined encoding) and replace them with byte values
2. Define the encoding of PHP scripts as ASCII, where bytes greater than
0x7F are allowed

alexanderholman commented:
As far as I was aware ASCII is limited to 7 bits? If that is the case then surely Unicode was correct and the definition needed updating to that?

Agares commented:
ASCII uses 7-bits, which in case of 8-bit bytes leaves the highest bit always set to 0. 
Unicode and UTF-8 are not synonyms. Unicode is basically a list of characters (+ specification which defines how they work together, how they are encoded, etc.). UTF-8 is one of unicode encodings, but it's not just "any bytes", but the unicode codepoints encoded in a specific way, in 1-4 bytes. The 0x80 to 0xFF codepoints actually encode as two bytes.

alexanderholman commented:
I understand that but ASCII is a defined set of characters, is it not? And I though the whole point of this PR was to reduce ambiguity? I don't see how that's really being achieved... Char 0x## could mean anything without a definition couldn't it? At least U+#### provided context. Random byte codes doesn't seem to me at least any less ambiguous.

But it's more than possible I just don't get it...

Agares commented:
U+#### was ambiguous as there are multiple Unicode encodings (UTF-7, UTF-8, UTF-16, UTF-32, none of them encode the ranges specified as a single byte).
PHP itself doesn't care about the exact byte values, and doesn't care about unicode. The sequence of 4 bytes: 0xFF 0xFF 0xFF 0xFF isn't valid in any Unicode encoding as far as I know (it certainly isn't valid in UTF-8), but it's perfectly valid as a name in PHP.

And defining the encoding as ASCII + additional byte values is needed, so things like keywords have a defined encoding.

nikic commented on spec/19-grammar.md:
```diff
@@ -3,6 +3,8 @@
 ## General
 
 The grammar notation is described in [Grammars section](09-lexical-structure.md#grammars).
+PHP scripts are encoded as ASCII, but bytes 0x80 to 0xFF are allowed in some places, as defined in the grammar. 
```
I think the precise way to specify this would be to say that PHP scripts are encoded using an ASCII-compatible encoding. In this case we can also keep the existing `U+00XX` notation, as for an ASCII-compatible encoding `U+00XX` for `0xXX < 0x80` will always be encoded as a `0xXX` byte.

Agares replied:
I think the `U+00XX` notation should be avoided, as this notation is usually used to refer to Unicode codepoints. 
If this is an ASCII-compatible encoding of Unicode, then how is `U+0100` encoded? 
And from different side - ISO-8859-2 is ASCII-compatible, but Unicode codepoints are meaningless in context of that encoding.

And sorry that I'm nitpicking on the details, but I believe that as a specification, this document should be as precise as it can be.

IMSoP replied:
Per my most recent comment on the Issue, I think a reasonable wording would be something like:

> PHP scripts are parsed as a series of 8-bit bytes, rather than code points from Unicode or any other character repertoire. Within this specification, bytes are represented by their ASCII interpretations where these are printable characters.

Agares replied:
I wouldn't go with "printable characters", as at least according to wikipedia everything below 0x20 isn't printable, which includes the tabulation character for example.

IMSoP replied:
I'm talking about characters that will be represented literally as ASCII characters in the specification; nobody's going to type a literal tab in the middle of a markdown file to represent the byte 0x09.

Agares replied:
Oh, right, my bad, seems like indeed the non-printable characters always have their byte-values specified.

IMSoP commented:
This is all completely the opposite of how PHP works. There is no requirement that source files be encoded in an ASCII-compatible encoding, nor is the grammar of the language defined in terms of bytes. See my comment under #233.

The only cases where we should talk about bytes is when discussing the contents of strings at run time, which are indeed arrays of bytes in no particular encoding.

IMSoP commented:
Previous comment retracted. My first claim, that there is no requirement for the source file to be ASCII-compatible, was correct; my second, that the grammar is expressed in Unicode was not. I'll carry on the discussion on #233

Agares commented:
So, it seems like we came to agreement. Is there anything I need to do before this can be merged?

nikic commented:
Sorry for the delay, I've merged this as https://github.com/php/php-langspec/commit/f35bd70531c012544e9061731415be8dcfe77f32 now.

I don't think this is entirely correct, but at least it's more correct than what we have now. Especially the statement that PHP scripts are ASCII encoded is not quite right. As I mentioned before, the technical requirement is that the encoding be ASCII-compatible, which is a term with a specific meaning: Unicode codepoints U+0000 to U+007F are encoded as bytes 0x00 through 0x7F, and bytes 0x00 through 0x7F do not appear as part of the encoding of any other codepoints.

This includes both UTF-8 and your example of ISO-8859-2. And it does make sense to talk about Unicode codepoints in the context of ISO-8859-2 -- while, this encoding cannot represent *all* Unicode codepoints, each character it represents also has an assigned Unicode codepoint. This is specified in auxiliary character set mapping tables published by the Unicode consortium, see ftp://ftp.unicode.org/Public/MAPPINGS/ISO8859/8859-2.TXT for this particular case.

nikic commented:
I've reverted this again in 9d99cf029b84f40b99bdab0224c79dd89c09e41a. I didn't notice that this changed the grammar file -- this is an auto-generated file and not an appropriate place for any spec text.

While looking for a good place to put this text instead I found https://github.com/php/php-langspec/blob/master/spec/09-lexical-structure.md#scripts, which requires that at least UTF-8 is supported and defines a dubious "transformation" step.

This is the section that needs to be adjusted, but probably in not quite the same way as just the added text in this PR.

Agares commented:
I'll look into  fixing that during the weekend. And by the way I think your previous comment was referring to an older version of the merge request, as the current one uses the wording that @IMSoP proposed, which mentions ASCII only as the representation used in the specification, but defines characters as bytes. Please have another look and I'm happy to address any concerns if you still have them, let's make this as correct as possible.

Agares commented:
Sorry for the delay. I have updated the MR and moved the text to "Lexical Structure" chapter. I also regenerated the grammar files with `php tools/grammar.php`.
Please review and thank you!

nikic commented:
Thanks for the update! Now merged as https://github.com/php/php-langspec/commit/fe55bee01293a925120cfe330b62a94fc5140e35.
